core/src/eval/mod.rs

```diff
@@ -960,7 +965,8 @@ pub fn subst<C: Cache>(
         | v @ Term::SealingKey(_)
         | v @ Term::Enum(_)
         | v @ Term::Import(_)
-        | v @ Term::ResolvedImport(_) => RichTerm::new(v, pos),
+        | v @ Term::ResolvedImport(_)
+        | v @ Term::Types(_) => RichTerm::new(v, pos),
```

Comment:
I wonder if this is the expected behavior. Morally, types can contain terms (imagine `Contract1 -> Contract2`), so we should very much continue the substitution by descending into the term, strictly speaking. This is the current behavior as `Contract1 -> Contract2` is turned into `$arrow Contract1 Contract2`, and thus is a normal term subject to substitution.

On the other hand, we don't substitute types/contracts in annotation. I suspect not substituting here is fine, because in the end putting  a type inside a variable means that you're going to use it as a contract anyway, and not say, serialize it, but at least this deserve a comment to say why we don't substitute under types.